Mathlib/FieldTheory/AxGrothendieck.lean

```diff
@@ -3,61 +3,248 @@ Copyright (c) 2023 Chris Hughes. All rights reserved.
+    (K ⊨ genericPolyMapSurjOnOfInjOn φ mons) ↔
+      ∀ (v : α → K) (p : { p : ι → MvPolynomial ι K // (∀ i, (p i).support ⊆ mons i) }),
+        let f : (ι → K) → (ι → K) := fun v i => eval v (p.1 i)
+        let S : Set (ι → K) := fun x => φ.Realize (Sum.elim v x)
+        S.MapsTo f S → S.InjOn f → S.SurjOn f S := by
+  letI := Classical.decEq K
+  letI := Classical.decEq ι
+  have injOnAlt : ∀ {S : Set (ι → K)} (f : (ι → K) → (ι → K)),
+      S.InjOn f ↔ ∀ x y, x ∈ S → y ∈ S → f x = f y → x = y := by
+    simp [Set.InjOn]; tauto
+  simp only [Sentence.Realize, Formula.Realize, genericPolyMapSurjOnOfInjOn, Formula.relabel,
+    Function.comp_def, Sum.map, id_eq, Equiv.sumAssoc, Equiv.coe_fn_symm_mk, Sum.elim_inr,
+    realize_iAlls, realize_imp, realize_relabel, Fin.natAdd_zero, realize_subst, realize_iInf,
+    Finset.mem_univ, realize_bdEqual, Term.realize_relabel, true_imp_iff,
+    Equiv.forall_congr_left (Equiv.curry (Fin 2) ι K), Equiv.curry_symm_apply, Function.uncurry,
+    Fin.forall_fin_succ_pi, Fin.forall_fin_zero_pi, realize_iExs, realize_inf, Sum.forall_sum,
+    Set.MapsTo, Set.mem_def, injOnAlt, funext_iff, Set.SurjOn, Set.image, setOf,
+    Set.subset_def, Equiv.forall_congr_left (mvPolynomialSupportLEEquiv mons)]
+  simp (config := { singlePass := true}) only [← Sum.elim_comp_inl_inr]
+  simp [Set.mem_def, Function.comp_def]
+
+theorem ACF_models_genericPolyMapSurjOnOfInjOn_of_prime [Fintype ι]
+    {p : ℕ} (hp : p.Prime) (φ : ring.Formula (α ⊕ ι)) (mons : ι → Finset (ι →₀ ℕ)) :
+    Theory.ACF p ⊨ᵇ genericPolyMapSurjOnOfInjOn φ mons := by
+  letI := Classical.decEq ι
+  haveI : Fact p.Prime := ⟨hp⟩
+  letI := compatibleRingOfRing (AlgebraicClosure (ZMod p))
+  haveI : CharP (AlgebraicClosure (ZMod p)) p :=
+    charP_of_injective_algebraMap
+      (RingHom.injective (algebraMap (ZMod p) (AlgebraicClosure (ZMod p)))) p
+  rw [← (ACF_isComplete (Or.inl hp)).realize_sentence_iff _
+    (AlgebraicClosure (ZMod p)), realize_genericPolyMapSurjOnOfInjOn]
+  rintro v ⟨f, _⟩
+  exact ax_grothendieck_of_locally_finite (K := ZMod p) (ι := ι) f _
+
+theorem ACF_models_genericPolyMapSurjOnOfInjOn_of_prime_or_zero
+    [Fintype ι] {p : ℕ} (hp : p.Prime ∨ p = 0)
+    (φ : ring.Formula (α ⊕ ι)) (mons : ι → Finset (ι →₀ ℕ)) :
+    Theory.ACF p ⊨ᵇ genericPolyMapSurjOnOfInjOn φ mons := by
+  rcases hp with hp | rfl
+  · exact ACF_models_genericPolyMapSurjOnOfInjOn_of_prime hp φ mons
+  · rw [ACF_zero_realize_iff_infinite_ACF_prime_realize]
+    convert Set.infinite_univ (α := Nat.Primes)
+    rw [Set.eq_univ_iff_forall]
+    intro ⟨p, hp⟩
+    exact ACF_models_genericPolyMapSurjOnOfInjOn_of_prime hp φ mons
+
+end FirstOrder
+
+open FirstOrder Language Field Ring MvPolynomial
+
+variable {K ι : Type*} [Field K] [IsAlgClosed K] [Finite ι]
+
+/-- A slight generalization of the **Ax-Grothendieck** theorem
+
+If `K` is an algebraically closed field, and `S` is a definable subset of `ι → K`, then any
```

Comment:
Here as well: I guess it is understood, but maybe mention that `ι` is finite.

Reply:
done